### PR TITLE
Look up installed tools from different toolshed(s)

### PR DESCRIPTION
**Pull request by mvdbeek**
On our servers we have a number of tools installed from a local toolshed and from the testtoolshed. Our users have been using these tools for a while, however we would like to be able to remove these tools and re-install them from the main toolshed (in the scope of them being available in main ...).

The problem is that if we do this, parameter choices will not be available anymore, workflows will claim that the tool is not available anymore and so on. I would be happy if this PR would be considered for inclusion.

With the PR applied, if finding a tool by its tool_id fails, we try again with the portion of the tool_id that corresponds to the tool shed replaced by the available tool sheds of the galaxy instance.

I made this behaviour a non-default setting controlled by lookup_tools_in_other_tool_sheds in galaxy.ini.sample

Feedback very welcome, as usual.

### CODE REVIEW

**Comment by jmchilton** [comment]
I am deeply ambivalent on this - in one sense I hate the ID resolution on the tool shed tools and I really want to simplify it and make it more concise, not make it more complicated. On the other hand, I really don't think there is a practical scenario where you wouldn't want to just enable this by default - if it goes in, it probably shouldn't even be a configuration option.

I'm going to leave this open for awhile, but I am tentatively a :+1: on it.

**Comment by hexylena** [comment]
Same opinion as @jmchilton, ambivalence. I always imagined this feature
would be realised as "tool upgrade paths" which would let people handle
moving tools around, deprecated tools pointing to a community maintained
version, and standard upgrades.... but that's a large amount of work and
still doesn't neatly handle your "upgrading tools in workflows" problem.
That's a large amount of work compared to hacking the tool ID resolution.

ons. 30. sep. 2015, 17.16 skrev John Chilton notifications@github.com:

**Comment by mvdbeek** [comment]
@jmchilton @erasche I absolutely agree, this would be very nice. If I can add something to the design: Would be awesome to go from a tool_id back to a github repo's changeset ...

If one day this is ready, we would still need to do some kind of migration. Do you think the proposed PR would make things more difficult down the road?

About the failing unit tests:
If I add `self.app.config.lookup_tools_in_other_tool_sheds = False` (or True) in test/unit/tools_support.py I can make the tests pass. Is this the way to go?

**Comment by jmchilton** [comment]
@mvdbeek I say wait another day, and if noone objects just drop the configuration option altogether. Just do this by default.

I wouldn't worry about repo to repo migrations at this point. It is not going to happen within the current tool shed framework I don't think, and if we do a radical redesign hopefully the IDs will be different enough in nature that this particular hack would not affect shed 2.0 tools.

File: lib/galaxy/config.py [hunk 1]
```diff
@@ -279,6 +279,7 @@ def __init__( self, **kwargs ):
         self.allow_library_path_paste = kwargs.get( 'allow_library_path_paste', False )
         self.disable_library_comptypes = kwargs.get( 'disable_library_comptypes', '' ).lower().split( ',' )
         self.watch_tools = kwargs.get( 'watch_tools', 'false' )
+        self.lookup_tools_in_other_tool_sheds = kwargs.get( 'lookup_tools_in_other_tool_sheds', 'false' )
```
Comment (nsoranzo) [inline lib/galaxy/config.py]: This should be:

```
        self.lookup_tools_in_other_tool_sheds = string_as_bool( kwargs.get( 'lookup_tools_in_other_tool_sheds', 'false' ) )
```

Comment (mvdbeek) [inline lib/galaxy/config.py]: OK, noted, but I'll do as @jmchilton said and remove the configurability, if there are no further objections.

File: lib/galaxy/tools/toolbox/base.py [hunk 2]
```diff
@@ -367,6 +368,40 @@ def _load_integrated_tool_panel_keys( self ):
...
+                        else:
+                            args = list(args)
+                            args[0] = alternate_tool_id
+                        tool = func(self, *args, **kwargs)
+                        if tool:
+                            return tool
+            return tool
+        return _manage_get_tool
+
+    @try_other_tool_sheds
```
Comment (nsoranzo) [inline lib/galaxy/tools/toolbox/base.py]: Why have you implemented this as a decorator instead of modifying `get_tool()` directly? Seems odd.

Comment (mvdbeek) [inline lib/galaxy/tools/toolbox/base.py]: Thinking about it there is no good reason. I'll modify `get_tool()` directly.

**Comment by mvdbeek** [comment]
@nsoranzo is this OK/better now?

**Comment by martenson** [comment]
I am okay with trying to resolve tools this way however I discussed this with @blankenberg and we think that a user-facing warning on the toolform is justified and needed.

edit: we do something similar when tool is being re-run and only newer vesion is available; you are given the new one and presented a warning

**Comment by mvdbeek** [comment]
@martenson Good point. Currently you will get a message that was aiming for a different version (e.g. `This job was initially run with tool id "lbcd41.snv.jussieu.fr/toolshed/repos/mvdbeek/sra_tools/sam_dump/1.2.4", version "1.2.4", which is currently not available. You can re-run the job with this tool, which is a derivation of the original tool.`) which is a bit weird. Should I see what can be done or is someone already on it? Perhaps explicitly mentionning the tool_id?

as such:
`This job was initially run with tool id "lbcd41.snv.jussieu.fr/toolshed/repos/mvdbeek/sra_tools/sam_dump/1.2.4", version "1.2.4", which is currently not available. You can re-run the job with this tool (testtoolshed/toolshed/repos/mvdbeek/sra_tools/sam_dump/1.2.4", version "1.2.4"), which is a derivation of the original tool.`

**Comment by martenson** [comment]
The tool_id is fine and should be included but given that it won't tell much to the majority of our users I would hope for something like `The tool you have requested is not available, you are presented a different tool which was automatically matched as similar to the requested one. It can produce different results. You can check details using the tool_id: ...` being shown in case the logic you implemented is triggered.

**Comment by jmchilton** [comment]
"This came from a different tool shed and is almost certainly the same tool but under some theoretical possible situations that would never occur in reality might not be."

This doesn't seem like a useful message for the end user and would be only confusing. If the admin does this that is the admins business and she or he is responsible, there is no reason for us to paternalisticly inject a confusing message about concerns the admin has decided her or his users do not need to worry about.

**Comment by nsoranzo** [comment]
@mvdbeek LGTM, thanks!

**Comment by martenson** [comment]
@jmchilton I think this is more serious and more probable situation than the 'here is the newer version because we do not have the old one' which does have the warning.
I believe we need both warnings but I am probably fine with removing them _and_ clearly explaining this behavior in admin docs i.e. here: https://wiki.galaxyproject.org/Admin/Tools

**Comment by mvdbeek** [comment]
I tend to agree with @jmchilton, but we can also have the situation that a user got a workflow that includes a tool for example from the testtoolshed from god knows where. In that situation the user should at least be aware that the tool doesn't match up 100%, right ?

**Comment by jmchilton** [comment]
@mvdbeek - No, users should never end up with a workflow with a tool from the test tool shed in it. It is not a corner case we should program logic for.

@martenson - If we don't believe these are different versions of the same tools we shouldn't merge this, but it is an unrealistic and purely theoretical concern - in practice there won't be two tool sheds with the same users and same tools but have those be different tools. There is not a tool shed out there with a devteam account and bwa tool that actually is bowtie.

**Comment by blankenberg** [comment]
There should be an user visible warning any time we cannot be 100% certain, which is the case when 'guessing' tools based upon partial ID matching.

**Comment by jmchilton** [comment]
Admins change installed tool wrappers all the time, it is not like we hash them. Also admins have access to the install database. It is not our job to get between admins and their users.

We don't make 100% guarantees about anything, we can't and we shouldn't. We should instead just use common sense.

**Comment by mvdbeek** [comment]
@jmchilton agreed, production workflows should not and probably rarely do include tools from the testtoolshed, so not a good example. But does this hold true also for community toolsheds such as http://toolshed.genouest.org/ or the gigascience toolshed? Or someone unfortunate enough that has to mirror a toolshed because he's behind a firewall? Maybe these example are too constructed, I admit that.

I agree that it is very unlikely that devteam bwa wouldn't be devteam bwa (that's the whole point of the PR), but I don't see the harm in telling the user that the tool origin has changed. The dangers is to confuse the user, but we could have a link pointing to the wiki that explains in detail why this warning message is being shown. Something like `Click to learn more`. I would be volunteering to write this.

In the greater picture I think it should be possible to reproduce analyses using workflows, even if they come from different galaxy instances, and these galaxy instances having their tools coming from different toolsheds. Because very likely they do the same thing. But we should be explicit about this.

I'm open for any suggestion.

**Comment by mvdbeek** [comment]
ping @jmchilton 
This last update is concerning the warning messages shown to users that rerun tools or workflows with tools whose tool_id has changed.
Attached are the various cases as screenshots

This is what users see if they click on the re-run button in the history.
screen shot 2015-12-05 at 10 20 01

This is what happens when they actually change the version.
screen shot 2015-12-05 at 10 19 22

And this is when you load a workflow with a tool that is being matched up.
screen shot 2015-12-05 at 12 18 53

**Comment by jmchilton** [comment]
Thanks for the changes... I guess on behalf of Dan who actually wanted them :smile:. There is a failing test now:

```
galaxy.web.framework.decorators: ERROR: Uncaught exception in exposed API method:
Traceback (most recent call last):
  File "/galaxy/lib/galaxy/web/framework/decorators.py", line 258, in decorator
    rval = func( self, trans, *args, **kwargs)
  File "/galaxy/lib/galaxy/webapps/galaxy/api/workflows.py", line 232, in workflow_dict
    ret_dict = self.workflow_contents_manager.workflow_to_dict( trans, stored_workflow, style=style )
  File "/galaxy/lib/galaxy/managers/workflows.py", line 277, in workflow_to_dict
    return self._workflow_to_dict_editor( trans, stored )
  File "/galaxy/lib/galaxy/managers/workflows.py", line 338, in _workflow_to_dict_editor
    elif module.version_changes:
AttributeError: 'InputDataModule' object has no attribute 'version_changes'
```

That test can be run locally with:

```
./run_tests.sh -api test/api/test_workflows.py:WorkflowsApiTestCase.test_export_editor
```

**Comment by mvdbeek** [comment]
@jmchilton thanks! I did fix the failing workflow test (The failing framework test is unrelated, right?), but it made me notice something here:
Neither unconfirmed tool_version nor tool_id changes cause the export to fail and/or display a user-facing message (This is pre-existing behaviour). It just gets silently converted to whatever is installed.

@blankenberg are the warning messages ok for you, and should we also address the the silent conversion when exporting workflows ?

**Comment by martenson** [comment]
The framework test is unrelated. I also trigerred a TS test https://jenkins.galaxyproject.org/job/docker-toolshed/20/console

I am :+1: on this, thanks @mvdbeek

**Comment by jmchilton** [comment]
I am -.1 on the warnings - but they are less intrusive to the code base than I assumed they would be, I remain :+1: on the PR overall. I'll give Dan a day to respond and then merge.

**Comment by bgruening** [comment]
@galaxybot test this

**Comment by hexylena** [comment]
:+1:

**Comment by bgruening** [comment]
@mvdbeek nice work! One question would it be possible to link the tool to the ToolShed from the warning?
This warning is needed I agree here with Dan but a little bit less disturbing would be nice. The ID is very technical so we could hide them under a pop-up or under a link.

@jmchilton don't forget that we made Galaxy so easy to install/administrate that there isn't always a full-time Galaxy admin. I think this warning does not hurt anyone and informs about a very crucial fact in reproducibility. Hopefully, this warning is not seen much.

**Comment by mvdbeek** [comment]
> @mvdbeek nice work! One question would it be possible to link the tool to the ToolShed from the warning?
> This warning is needed I agree here with Dan but a little bit less disturbing would be nice. The ID is very technical so we could hide them under a pop-up or under a link.

Yes, that will work, except that for uninstalled tools we don't have a changeset revision (this should be stored in the workflow, imo), so the best we can do is to link to the repo for the old tool, and for the new tool we can link to the installed changeset revision.

**Comment by bgruening** [comment]
Would be fine with me. Still wondering why the revision is not part of the WF, I thought it is. This seems like a bug.

**Comment by mvdbeek** [comment]
I might be wrong, but I think this is only filled in during invocation.
Easy to check though; *.ga files don't contain changeset revisions.

**Comment by jmchilton** [comment]
@bgruening I frequently forget "we made Galaxy so easy to install/administrate" - I forget this fact like all the time - every second :smile:.

**Comment by martenson** [comment]
thanks @mvdbeek !
